io.openems.shared.mssql/src/io/openems/shared/mssql/MssqlConnector.java

```diff
@@ -0,0 +1,173 @@
+	 * @param dbname				Database name
+	 * @param port					Port number in which the MS-SQL DB is available
+	 * @param username				User name for logging in
+	 * @param password				Password for logging in
+	 * @param isPasswordProtected	Is the DB password protected? then the inputs are handled differently
+	 * @param isReadOnly			Is the database supposed to be read only?, then writing would be skipped.
+	 */
+	public MssqlConnector(String server, String dbname, int port, String username, String password, 
+			boolean isPasswordProtected, boolean isReadOnly) {
+		super();
+		this.server = server;
+		this.dbname = dbname;
+		this.port = port;
+		this.username = username;
+		this.password = password;
+		this.isPasswordProtected = isPasswordProtected;
+		this.isReadOnly = isReadOnly;
+		this.debugLogExecutor.scheduleWithFixedDelay(() -> {
+			int queueSize = this.executor.getQueue().size();
+			this.log.info(String.format("[monitor] Pool: %d, Active: %d, Pending: %d, Completed: %d %s",
+					this.executor.getPoolSize(), //
+					this.executor.getActiveCount(), //
+					this.executor.getQueue().size(), //
+					this.executor.getCompletedTaskCount(), //
+					(queueSize == EXECUTOR_QUEUE_SIZE) ? "!!!BACKPRESSURE!!!" : "")); //
+		}, 10, 10, TimeUnit.SECONDS);
+	}
+	
+	private Connection _mssqlDB = null;
+	
+	/**
+	 * Get the MS-SQL Connection
+	 * @return	connection
+	 */
+	private Connection getConnection() {
+		if (this._mssqlDB == null) {
+			SQLServerDataSource MssqlDB = new SQLServerDataSource();
+			// Source:
+			// https://docs.microsoft.com/de-de/sql/connect/jdbc/data-source-sample?view=sql-server-ver15
+			MssqlDB.setServerName(this.server);
+			MssqlDB.setDatabaseName(this.dbname);
+			MssqlDB.setPortNumber(this.port);
+			if (this.isPasswordProtected) {
+				MssqlDB.setUser(this.username);
+				MssqlDB.setPassword(this.password);
+			}
+			try {
+				this._mssqlDB = MssqlDB.getConnection();
+			} catch (SQLServerException e) {
+				log.warn("Unable to establish connection with MS-SQL DB. Exception: "+e.getMessage());
+			}
+		}
+		return this._mssqlDB;
+	}
+	
+	public void deactivate() {
+		// shutdown executor
+		if (this.executor != null) {
+			try {
+				this.executor.shutdown();
+				this.executor.awaitTermination(5, TimeUnit.SECONDS);
+			} catch (InterruptedException e) {
+				this.log.warn("tasks interrupted");
+			}finally {
+				if (!this.executor.isTerminated()) {
+					this.log.warn("cancel non-finished tasks");
+				}
+				this.executor.shutdownNow();
+			}
+		}
+		if (this._mssqlDB != null) {
+			try {
+				this._mssqlDB.close();
+			} catch (SQLException e) {
+				log.warn("Unable to close MS-SQL DB. Exception: " + e.getMessage());
+			}
+		}
+	}
+	
+	/**CODE TO BE CHECKED
+	public void write(String query) throws OpenemsException {
+		if (this.isReadOnly) {
+			log.info("Read-Only-Mode is activated. Not writing measurement: " +
+					StringUtils.toShortString(query, 100));
+		}
+		try {
```

Comment:
```
this.executor.execute(() -> {
	try {
		this.getConnection().createStatement().executeUpdate(query);
		this.getConnection().commit();
	} catch (SQLException e) {
		this.log.warn("Unable to write point: " + e.getMessage());
		try {
			this.getConnection().rollback();
		} catch (SQLException e1) {
			e1.printStackTrace();
		}
	}
});
```

Reply:
Hi Laksh. If you are not used to Java Lamda functions this might have been a bit tricky to understand. The part inside the execute()-method call `() -> {...}` is actually called asynchronously by the executor. Because of this you cannot simply throw the Exception the way you did.

All in all your solution looks good, but is maybe to complicated for the beginning. The structure with a 'shared' bundle for InfluxDB is only there to make the interface available for Backend and Edge. From what I understood this might not (yet) be required in your case, so the architecture could be done simpler. 

Anyhow, the above commented code should work for you. Only downside is, that you will not get an Exception thrown on write errors which you could act on. Right now you would just get a warning on the console log.

Regards,
Stefan

Reply:
Hi Stefan, 
Thanks for the quick review. I have checked it, and the code works as per your suggestion. I am new to JAVA, and yes Lambda functions are not quiet clear to me yet. I appreciate your help. 

Also, yes you are right, currently my requirment is only to write the values from the components to the DB. But in future I also plan in integrating Forecast algorithms (which will carry out the forecasting and write data to DB), which as per my understanding is a part of the backend. Along with that I believe there would be instances where my algorithms (also as per my understanding, a part of the backend) would have to directly interact with the DB (both read and write). In such cases, I think connection elements in a shared file would be necessary. Please correct me if I am wrong. 

Thanks and regards
Laksh

Reply:
Hi Laksh,
OpenEMS really tries to solve most of the control part locally, so that a Backend should not strictly be required for any control logic. Interaction with a `Timedata` service for reading and writing locally is absolutely possible. So keeping everything in the "OpenEMS Edge"-namespace for now would be sufficient and keep things simple. But your approach is not wrong, of course.

Also again, OpenEMS tries to be as modular and 'pluggable'  as possible. See for example the simple implementation of a 'Persistence-Model-Predictor', that simply 'predicts' the next day with the values from last day. Technically in OpenEMS there is:
- The `Predictor`-Api defines what a Predictor-Service needs to provide, i.e. a `get24HoursPrediction()` method (see https://github.com/OpenEMS/openems/blob/develop/io.openems.edge.predictor.api/src/io/openems/edge/predictor/api/oneday/Predictor24Hours.java)
- The `Timedata`-Api defines what a Timedata-Service needs to provide, e.g. a `queryHistoricData()` method. (see https://github.com/OpenEMS/openems/blob/develop/io.openems.edge.timedata.api/src/io/openems/edge/timedata/api/Timedata.java and https://github.com/OpenEMS/openems/blob/e4ef05bf41f0378c75ece9e9bf435e0b3483281b/io.openems.common/src/io/openems/common/timedata/CommonTimedataService.java)
- `Predictor.PersistenceModel` uses a `Timedata` Service to query the values for the last 24 hours and uses them to serve a Prediction. (see https://github.com/OpenEMS/openems/blob/develop/io.openems.edge.predictor.persistencemodel/src/io/openems/edge/predictor/persistencemodel/PersistenceModelPredictorImpl.java#L75)

This modular structure allows, e.g.
- to use your MSSQL-Timedata-Service immediately for this simple prediction
- to write advanced predection algorithms that use any timedata service, e.g. to feed a neural network
- or you could at any time implement a custom predictor, that uses a webservice (see #1422)